GitUI/BranchTreePanel/RepoObjectsTree.SettingsContextMenu.cs

```diff
@@ -1,30 +1,119 @@
 using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Windows.Forms;
 using GitCommands;
 
 namespace GitUI.BranchTreePanel
 {
     public partial class RepoObjectsTree
     {
+        private Dictionary<Tree, int> GetTreeToPositionIndex()
+        {
+            return new Dictionary<Tree, int>
+            {
+                [_branchesTree] = AppSettings.RepoObjectsTreeBranchesIndex,
+                [_remotesTree] = AppSettings.RepoObjectsTreeRemotesIndex,
+                [_tagTree] = AppSettings.RepoObjectsTreeTagsIndex,
+                [_submoduleTree] = AppSettings.RepoObjectsTreeSubmodulesIndex
+            };
+        }
+
+        private void SaveTreeToPositionIndex(Dictionary<Tree, int> treeToPositionIndex)
+        {
+            AppSettings.RepoObjectsTreeBranchesIndex = treeToPositionIndex[_branchesTree];
+            AppSettings.RepoObjectsTreeRemotesIndex = treeToPositionIndex[_remotesTree];
+            AppSettings.RepoObjectsTreeTagsIndex = treeToPositionIndex[_tagTree];
+            AppSettings.RepoObjectsTreeSubmodulesIndex = treeToPositionIndex[_submoduleTree];
+        }
+
+        private void RebuildMenuSettings()
+        {
+            var indexToTree = GetTreeToPositionIndex().ToDictionary(kvp => kvp.Value, kvp => kvp.Key);
+
+            treeMain.BeginUpdate();
+
+            // For now, the only items in menuSettings are the "show" items, so just clear it all,
+            // and re-add them in current index order.
+            menuSettings.Items.Clear();
+            var treeToShowMenuItem = new Dictionary<Tree, ToolStripMenuItem>
+            {
+                [_branchesTree] = tsmiShowBranches,
+                [_remotesTree] = tsmiShowRemotes,
+                [_tagTree] = tsmiShowTags,
+                [_submoduleTree] = tsmiShowSubmodules
+            };
+            for (int i = 0; i < treeToShowMenuItem.Count(); ++i)
+            {
+                var tsmi = treeToShowMenuItem[indexToTree[i]];
```

Comment:
This will not chandle corrupted indexes (for instance if adding a new tree and downgrading)